File: src/compiler.js
import { readFile, writeFile, readdir, stat } from 'node:fs/promises';
import { resolve, join, extname, basename } from 'node:path';
import { colors, ui, Spinner } from './ui.js';
import { notifyClients } from './server.js';

const API_BASE = 'https://latex.taptapp.xyz';

async function getFiles(dir, baseDir = dir) {
    const entries = await readdir(dir, { withFileTypes: true });
    const files = [];
    for (const entry of entries) {
        const res = join(dir, entry.name);
        if (entry.isDirectory()) {
            if (entry.name !== 'node_modules' && !entry.name.startsWith('.')) {
                files.push(...(await getFiles(res, baseDir)));
            }
        } else {
            const rel = join(dir.replace(baseDir, ''), entry.name).replace(/^[\\\/]/, '').replace(/\\/g, '/');
            files.push({
                path: res,
                relative: rel
            });
        }
    }
    return files;
}

export async function compile(dir, options) {
    const outputFileName = options.output || 'output.pdf';
    const serverUrl = options.server || API_BASE;
    const spinner = new Spinner(`${colors.blue}Preparing compilation...`).start();

    try {
        const absoluteDir = resolve(dir);
        const allFiles = await getFiles(absoluteDir);

        // Sort files to prioritize .tex files (upladed first)
        allFiles.sort((a, b) => {
            const extA = extname(a.path).toLowerCase();
            const extB = extname(b.path).toLowerCase();
            if (extA === '.tex' && extB !== '.tex') return -1;
            if (extA !== '.tex' && extB === '.tex') return 1;
            return a.relative.localeCompare(b.relative);
        });

        const form = new FormData();
        const validateForm = new FormData();
        let hasLatex = false;
        let sentFilesCount = 0;

        for (const fileObj of allFiles) {
            const ext = extname(fileObj.path).toLowerCase();
            const allowedExts = ['.tex', '.bib', '.sty', '.cls', '.pdf', '.png', '.jpg', '.jpeg', '.csv'];

            if (allowedExts.includes(ext)) {
                // Skip the output file to avoid re-uploading it in subsequent runs
                if (basename(fileObj.path) === outputFileName) continue;

                const content = await readFile(fileObj.path);
                const blob = new Blob([content]);
                form.append('files', blob, fileObj.relative);
                sentFilesCount++;

                if (ext === '.tex') {
                    validateForm.append('files', blob, fileObj.relative);
                    hasLatex = true;
                }
            }
        }

        if (!hasLatex) {
            spinner.fail(`${colors.yellow}No LaTeX files found in directory.`);
            return;
        }

        // Phase 1: Validation
        spinner.update(`${colors.cyan}Validating LaTeX...`);
        try {
            const validateRes = await fetch(`${serverUrl}/validate`, {
                method: 'POST',
                body: validateForm
            });
            const validation = await validateRes.json();

            if (validation.valid === false) {
                spinner.stop();
                ui.error('Validation failed!');
                validation.errors.forEach(err => {
                    console.log(`  ${colors.red}• [Line ${err.line || '?'}] ${err.message}${colors.reset}`);
                });
                return;
            }

            if (validation.warnings?.length > 0) {
                spinner.stop();
                ui.warn('Validation warnings:');
                validation.warnings.forEach(warn => console.log(`  ${colors.yellow}⚡ ${warn}${colors.reset}`));
                spinner.start();
            }
        } catch (e) {
            // Validation service unavailable, proceed anyway
        }

        // Phase 2: Compilation
        spinner.update(`${colors.cyan}Compiling via Tachyon-Tex API...`);

        const response = await fetch(`${serverUrl}/compile`, {
            method: 'POST',
            body: form
        });

        if (response.ok) {
            const buffer = await response.arrayBuffer();
            const compileTime = response.headers.get('x-compile-time-ms') || 'unknown';
            const filesReceived = response.headers.get('x-files-received') || sentFilesCount;

            const outputPath = resolve(process.cwd(), dir, outputFileName);
            await writeFile(outputPath, Buffer.from(buffer));

            if (options.watch) {
                notifyClients(outputPath);
            }

            spinner.succeed(`${colors.green}PDF generated in ${colors.bold}${compileTime}ms${colors.reset}`);
            console.log(`${colors.dim}  ⚡ Files: ${filesReceived} processed${colors.reset}`);
            console.log(`${colors.dim}  📍 Path: ${outputPath}${colors.reset}`);
        } else {
            const errorMsg = await response.text();
            spinner.fail(`Compilation failed (Status ${response.status})`);
            console.log(`\n${colors.yellow}--- Error Log ---${colors.reset}`);
            console.log(`${colors.red}${errorMsg || 'Error details unavailable.'}${colors.reset}`);
            console.log(`${colors.yellow}-----------------\n${colors.reset}`);
        }
    } catch (error) {
        spinner.fail('An unexpected error occurred.');
        console.error(`${colors.red}${error.message}${colors.reset}`);
    }
}

File: src/compiler.rs
use std::path::Path;
use std::fs;
use reqwest::multipart::{Form, Part};
use crate::ui::Spinner;
use colored::*;

const API_BASE: &str = "https://latex.taptapp.xyz";

const ALLOWED_EXTS: [&str; 17] = [
    "tex", "bib", "sty", "cls", "bst",
    "pdf", "png", "jpg", "jpeg", "eps",
    "csv", "dat", "tsv", "txt",
    "tikz", "otf", "ttf"
];

pub async fn compile(dir: &Path, output_name: &str, silent: bool) -> Result<(), Box<dyn std::error::Error>> {
    let spinner = if silent { None } else { Some(Spinner::new("Compiling...")) };

    let absolute_dir = fs::canonicalize(dir)?;
    let mut form = Form::new();
    let mut file_count = 0;

    for entry in walkdir::WalkDir::new(&absolute_dir)
        .into_iter()
        .filter_map(|e| e.ok())
        .filter(|e| e.file_type().is_file())
    {
        let path = entry.path();
        let ext = path.extension().and_then(|s| s.to_str()).unwrap_or("").to_lowercase();
        
        if !ALLOWED_EXTS.contains(&ext.as_str()) {
            continue;
        }

        // Skip output file
        if path.file_name().and_then(|s| s.to_str()) == Some(output_name) {
            continue;
        }

        let relative = path.strip_prefix(&absolute_dir)?.to_string_lossy().replace('\\', "/");
        let content = fs::read(path)?;
        
        form = form.part("files", Part::bytes(content).file_name(relative));
        file_count += 1;
    }

    if file_count == 0 {
        if let Some(s) = spinner { s.fail("No files found"); }
        return Err("No LaTeX files found".into());
    }

    if let Some(ref s) = spinner { s.update("Sending to server..."); }

    let client = reqwest::Client::new();
    let response = client
        .post(format!("{}/compile", API_BASE))
        .multipart(form)
        .send()
        .await?;

    if response.status().is_success() {
        let compile_time: String = response.headers()
            .get("x-compile-time-ms")
            .and_then(|h| h.to_str().ok())
            .unwrap_or("?")
            .to_string();

        let bytes = response.bytes().await?;
        let output_path = dir.join(output_name);
        fs::write(&output_path, bytes)?;

        if let Some(s) = spinner {
            s.succeed(&format!("Done in {}ms → {}", compile_time.bold(), output_path.display()));
        }

        crate::server::notify_clients(&output_path).await;
        Ok(())
    } else {
        let err = response.text().await?;
        if let Some(s) = spinner { s.fail("Compilation failed"); }
        Err(err.into())
    }
}

pub async fn verify(dir: &Path, silent: bool) -> Result<(), Box<dyn std::error::Error>> {
    let spinner = if silent { None } else { Some(Spinner::new("Validating...")) };

    let absolute_dir = fs::canonicalize(dir)?;
    let mut form = Form::new();
    let mut file_count = 0;

    for entry in walkdir::WalkDir::new(&absolute_dir)
        .into_iter()
        .filter_map(|e| e.ok())
        .filter(|e| e.file_type().is_file())
    {
        let path = entry.path();
        let ext = path.extension().and_then(|s| s.to_str()).unwrap_or("").to_lowercase();
        
        if ext != "tex" {
            continue;
        }

        let relative = path.strip_prefix(&absolute_dir)?.to_string_lossy().replace('\\', "/");
        let content = fs::read(path)?;
        
        form = form.part("files", Part::bytes(content).file_name(relative));
        file_count += 1;
    }

    if file_count == 0 {
        if let Some(s) = spinner { s.fail("No .tex files found"); }
        return Err("No LaTeX files found".into());
    }

    if let Some(ref s) = spinner { s.update("Sending to server..."); }

    let client = reqwest::Client::new();
    let response = client
        .post(format!("{}/validate", API_BASE))
        .multipart(form)
        .send()
        .await?;

    if response.status().is_success() {
        #[derive(serde::Deserialize)]
        struct ValidationResponse {
            valid: bool,
            errors: Option<Vec<ValidationError>>,
        }
        #[derive(serde::Deserialize)]
        struct ValidationError {
            line: Option<u32>,
            message: String,
        }

        let result: ValidationResponse = response.json().await?;
        
        if result.valid {
            if let Some(s) = spinner { s.succeed("Validation passed ✓"); }
            Ok(())
        } else {
            if let Some(s) = spinner { s.fail("Validation failed"); }
            let errors = result.errors.unwrap_or_default();
            let msg = errors.iter()
                .map(|e| format!("[Line {:?}] {}", e.line, e.message))
                .collect::<Vec<_>>()
                .join("\n");
            Err(msg.into())
        }
    } else {
        let err = response.text().await?;
        if let Some(s) = spinner { s.fail("Validation request failed"); }
        Err(err.into())
    }
}
